frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,907 @@
+	/// [`QueuedSolution`].
+	pub fn finalize_signed_phase() -> (bool, Weight) {
+		let mut all_submissions = Self::signed_submissions();
+		let mut found_solution = false;
+		let mut weight = T::DbWeight::get().reads(1);
+
+		let SolutionOrSnapshotSize { voters, targets } =
+			Self::snapshot_metadata().unwrap_or_default();
+
+		while let Some(best) = all_submissions.pop_last() {
+			let SignedSubmission { solution, who, deposit, reward } = best;
+			let active_voters = solution.compact.voter_count() as u32;
+			let feasibility_weight = {
+				// defensive only: at the end of signed phase, snapshot will exits.
+				let desired_targets = Self::desired_targets().unwrap_or_default();
+				T::WeightInfo::feasibility_check(
+					voters,
+					targets,
+					active_voters,
+					desired_targets,
+				)
+			};
+			// the feasibility check itself has some weight
+			weight = weight.saturating_add(feasibility_weight);
+			match Self::feasibility_check(solution, ElectionCompute::Signed) {
+				Ok(ready_solution) => {
+					Self::finalize_signed_phase_accept_solution(
+						ready_solution,
+						&who,
+						deposit,
+						reward,
+					);
+					found_solution = true;
+
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_accept_solution());
+					break;
+				}
+				Err(_) => {
+					Self::finalize_signed_phase_reject_solution(&who, deposit);
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_reject_solution());
+				}
+			}
+		}
+
+		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
+		// they didn't end up being used. Unreserve the bonds.
+		let discarded = all_submissions.len();
+		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
+			let _remaining = T::Currency::unreserve(&who, deposit);
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			debug_assert!(_remaining.is_zero());
+		}
+
+		debug_assert!(!SignedSubmissionIndices::<T>::exists());
+		debug_assert!(!SignedSubmissionNextIndex::<T>::exists());
+		debug_assert!(SignedSubmissionsMap::<T>::iter().next().is_none());
+
+		log!(debug, "closed signed phase, found solution? {}, discarded {}", found_solution, discarded);
+		(found_solution, weight)
+	}
+
+	/// Helper function for the case where a solution is accepted in the signed phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_accept_solution(
+		ready_solution: ReadySolution<T::AccountId>,
+		who: &T::AccountId,
+		deposit: BalanceOf<T>,
+		reward: BalanceOf<T>,
+	) {
+		// write this ready solution.
+		<QueuedSolution<T>>::put(ready_solution);
+
+		// unreserve deposit.
+		let _remaining = T::Currency::unreserve(who, deposit);
+		debug_assert!(_remaining.is_zero());
+
+		// Reward.
+		let positive_imbalance = T::Currency::deposit_creating(who, reward);
+		T::RewardHandler::on_unbalanced(positive_imbalance);
```

Comment:
Should we use `Rewarded` event here? Or is it confusing if handler just burns the funds?

Reply:
Pretty good point, this and `Slashed` must be used.

Reply:
While the info can be deduced off chain using storage queries, I think having these events will be nice QoL for looking things up in event indexing block explorers (e.g. subscan)

Reply:
You're right: bf4572c96.